client/clientimpl.go

```diff
@@ -0,0 +1,348 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gorilla/websocket"
+
+	"github.com/open-telemetry/opamp-go/client/internal"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+// Impl is a Client implementation.
+type Impl struct {
+	logger   types.Logger
+	settings StartSettings
+
+	// OpAMP Server URL.
+	url *url.URL
+
+	// HTTP request headers to use when connecting to OpAMP Server.
+	requestHeader http.Header
+
+	// Websocket dialer and connection.
+	dialer    websocket.Dialer
+	conn      *websocket.Conn
+	connMutex sync.RWMutex
+
+	// True is Start() is successful.
+	isStarted bool
+
+	// Context to use for background running go routines.
+	runContext context.Context
+	runCancel  context.CancelFunc
+
+	// True when stopping is in progress.
+	isStoppingFlag  bool
+	isStoppingMutex sync.RWMutex
+
+	// Indicates that the Client is fully stopped.
+	stoppedSignal chan struct{}
+
+	// The sender is responsible for sending portion of the OpAMP protocol.
+	sender *internal.Sender
+}
+
+var _ OpAMPClient = (*Impl)(nil)
+
+const retryAfterHTTPHeader = "Retry-After"
+
+func New(logger types.Logger) *Impl {
+	if logger == nil {
+		logger = &nopLogger{}
+	}
+
+	w := &Impl{
+		logger:        logger,
+		stoppedSignal: make(chan struct{}, 1),
+		sender:        internal.NewSender(logger),
+	}
+	return w
+}
+
+func (w *Impl) Start(settings StartSettings) error {
+	if w.isStarted {
+		return errors.New("already started")
```

Comment:
This is really a nitty nit, but perhaps all errors could be defined at the top of the file and referenced later?

Reply:
I defined unexported errors at the top.